{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This notebook was created to assess convergent evolution (i.e. homoplasy score) for mutations in the loci interrogated for this study."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:100% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:100% !important; }</style>\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sys\n",
    "from collections import Counter\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "from matplotlib import gridspec"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# [1] Load in Homoplasy Scores from *SNPPar* for SNPs & *TopDis (spacer length = 4)* for INDELs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### This data was loaded from another study assessing parallel evolution (i.e. homoplasy score) for all mutations detected across the genome in 31,428 isolates: \"Phase variation as a major mechanism of adaptation in *Mycobacterium tuberculosis* complex\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load in SNPs\n",
    "all_homoplasic_SNPs = pd.read_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/CSV_files/homoplasies_detected_in_global_lineages/SNP homoplasies from SNPPar/homoplasy_count_across_lineages.pkl')\n",
    "\n",
    "#load in INDELs\n",
    "all_homoplasic_INDELs = pd.read_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/CSV_files/homoplasies_detected_in_global_lineages/INDEL homoplasies/homoplasy_count_across_lineages_spacer_4.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>minor</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_pos</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>SNP_type</th>\n",
       "      <th>AA_change</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4A</th>\n",
       "      <th>4B</th>\n",
       "      <th>4C</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>total</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>num_isolates_alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2155168.0</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv1908c</td>\n",
       "      <td>944</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>N</td>\n",
       "      <td>S315T</td>\n",
       "      <td>203</td>\n",
       "      <td>408</td>\n",
       "      <td>297</td>\n",
       "      <td>306</td>\n",
       "      <td>310</td>\n",
       "      <td>179</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>1717</td>\n",
       "      <td>katG</td>\n",
       "      <td>8066</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>761155.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>1349</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>N</td>\n",
       "      <td>S450L</td>\n",
       "      <td>75</td>\n",
       "      <td>415</td>\n",
       "      <td>199</td>\n",
       "      <td>222</td>\n",
       "      <td>311</td>\n",
       "      <td>173</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>1406</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>5706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>781687.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0682</td>\n",
       "      <td>128</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>N</td>\n",
       "      <td>K43R</td>\n",
       "      <td>37</td>\n",
       "      <td>463</td>\n",
       "      <td>102</td>\n",
       "      <td>65</td>\n",
       "      <td>53</td>\n",
       "      <td>51</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>779</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>3509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4247429.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv3795</td>\n",
       "      <td>916</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>N</td>\n",
       "      <td>M306V</td>\n",
       "      <td>32</td>\n",
       "      <td>406</td>\n",
       "      <td>67</td>\n",
       "      <td>76</td>\n",
       "      <td>99</td>\n",
       "      <td>60</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>743</td>\n",
       "      <td>embB</td>\n",
       "      <td>2102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7582.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>281</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>N</td>\n",
       "      <td>D94G</td>\n",
       "      <td>10</td>\n",
       "      <td>320</td>\n",
       "      <td>67</td>\n",
       "      <td>77</td>\n",
       "      <td>53</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>552</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>844</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         pos ref alt minor  gene_id gene_pos          gene_category SNP_type  \\\n",
       "0  2155168.0   C   G     G  Rv1908c      944  Antibiotic Resistance        N   \n",
       "1   761155.0   C   T     T   Rv0667     1349  Antibiotic Resistance        N   \n",
       "2   781687.0   A   G     G   Rv0682      128  Antibiotic Resistance        N   \n",
       "4  4247429.0   A   G     G   Rv3795      916  Antibiotic Resistance        N   \n",
       "6     7582.0   A   G     G   Rv0006      281  Antibiotic Resistance        N   \n",
       "\n",
       "  AA_change    1    2    3   4A   4B   4C  5  6  total gene_name  \\\n",
       "0     S315T  203  408  297  306  310  179  8  6   1717      katG   \n",
       "1     S450L   75  415  199  222  311  173  8  3   1406      rpoB   \n",
       "2      K43R   37  463  102   65   53   51  6  2    779      rpsL   \n",
       "4     M306V   32  406   67   76   99   60  2  1    743      embB   \n",
       "6      D94G   10  320   67   77   53   25  0  0    552      gyrA   \n",
       "\n",
       "   num_isolates_alt  \n",
       "0              8066  \n",
       "1              5706  \n",
       "2              3509  \n",
       "4              2102  \n",
       "6               844  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_homoplasic_SNPs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(834981, 20)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(all_homoplasic_SNPs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "create **key** column for SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_homoplasic_SNPs_key_col = [ref + '_' + str(int(pos)) + '_' + alt for ref,pos,alt in zip(all_homoplasic_SNPs.ref,all_homoplasic_SNPs.pos,all_homoplasic_SNPs.alt)]\n",
    "all_homoplasic_SNPs.loc[:,'key'] = all_homoplasic_SNPs_key_col"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>minor</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_pos</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>SNP_type</th>\n",
       "      <th>AA_change</th>\n",
       "      <th>1</th>\n",
       "      <th>...</th>\n",
       "      <th>3</th>\n",
       "      <th>4A</th>\n",
       "      <th>4B</th>\n",
       "      <th>4C</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>total</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>num_isolates_alt</th>\n",
       "      <th>key</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2155168.0</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv1908c</td>\n",
       "      <td>944</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>N</td>\n",
       "      <td>S315T</td>\n",
       "      <td>203</td>\n",
       "      <td>...</td>\n",
       "      <td>297</td>\n",
       "      <td>306</td>\n",
       "      <td>310</td>\n",
       "      <td>179</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>1717</td>\n",
       "      <td>katG</td>\n",
       "      <td>8066</td>\n",
       "      <td>C_2155168_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>761155.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>1349</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>N</td>\n",
       "      <td>S450L</td>\n",
       "      <td>75</td>\n",
       "      <td>...</td>\n",
       "      <td>199</td>\n",
       "      <td>222</td>\n",
       "      <td>311</td>\n",
       "      <td>173</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>1406</td>\n",
       "      <td>rpoB</td>\n",
       "      <td>5706</td>\n",
       "      <td>C_761155_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>781687.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0682</td>\n",
       "      <td>128</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>N</td>\n",
       "      <td>K43R</td>\n",
       "      <td>37</td>\n",
       "      <td>...</td>\n",
       "      <td>102</td>\n",
       "      <td>65</td>\n",
       "      <td>53</td>\n",
       "      <td>51</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>779</td>\n",
       "      <td>rpsL</td>\n",
       "      <td>3509</td>\n",
       "      <td>A_781687_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4247429.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv3795</td>\n",
       "      <td>916</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>N</td>\n",
       "      <td>M306V</td>\n",
       "      <td>32</td>\n",
       "      <td>...</td>\n",
       "      <td>67</td>\n",
       "      <td>76</td>\n",
       "      <td>99</td>\n",
       "      <td>60</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>743</td>\n",
       "      <td>embB</td>\n",
       "      <td>2102</td>\n",
       "      <td>A_4247429_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7582.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>281</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>N</td>\n",
       "      <td>D94G</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>67</td>\n",
       "      <td>77</td>\n",
       "      <td>53</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>552</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>844</td>\n",
       "      <td>A_7582_G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         pos ref alt minor  gene_id gene_pos          gene_category SNP_type  \\\n",
       "0  2155168.0   C   G     G  Rv1908c      944  Antibiotic Resistance        N   \n",
       "1   761155.0   C   T     T   Rv0667     1349  Antibiotic Resistance        N   \n",
       "2   781687.0   A   G     G   Rv0682      128  Antibiotic Resistance        N   \n",
       "4  4247429.0   A   G     G   Rv3795      916  Antibiotic Resistance        N   \n",
       "6     7582.0   A   G     G   Rv0006      281  Antibiotic Resistance        N   \n",
       "\n",
       "  AA_change    1     ...         3   4A   4B   4C  5  6  total  gene_name  \\\n",
       "0     S315T  203     ...       297  306  310  179  8  6   1717       katG   \n",
       "1     S450L   75     ...       199  222  311  173  8  3   1406       rpoB   \n",
       "2      K43R   37     ...       102   65   53   51  6  2    779       rpsL   \n",
       "4     M306V   32     ...        67   76   99   60  2  1    743       embB   \n",
       "6      D94G   10     ...        67   77   53   25  0  0    552       gyrA   \n",
       "\n",
       "  num_isolates_alt          key  \n",
       "0             8066  C_2155168_G  \n",
       "1             5706   C_761155_T  \n",
       "2             3509   A_781687_G  \n",
       "4             2102  A_4247429_G  \n",
       "6              844     A_7582_G  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_homoplasic_SNPs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNP_variant_keys = set(all_homoplasic_SNPs.key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "834961"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(SNP_variant_keys)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### INDELs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_pos</th>\n",
       "      <th>ins_del</th>\n",
       "      <th>INDEL_type</th>\n",
       "      <th>codon_pos</th>\n",
       "      <th>...</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4A</th>\n",
       "      <th>4B</th>\n",
       "      <th>4C</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>total</th>\n",
       "      <th>num_isolates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50204</th>\n",
       "      <td>4408100</td>\n",
       "      <td>GC</td>\n",
       "      <td>G</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gid</td>\n",
       "      <td>Rv3919c</td>\n",
       "      <td>103</td>\n",
       "      <td>del</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>35</td>\n",
       "      <td>...</td>\n",
       "      <td>41.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>202.0</td>\n",
       "      <td>388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50198</th>\n",
       "      <td>4408087</td>\n",
       "      <td>CG</td>\n",
       "      <td>C</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gid</td>\n",
       "      <td>Rv3919c</td>\n",
       "      <td>116</td>\n",
       "      <td>del</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>39</td>\n",
       "      <td>...</td>\n",
       "      <td>23.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>130.0</td>\n",
       "      <td>212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50144</th>\n",
       "      <td>4407851</td>\n",
       "      <td>GC</td>\n",
       "      <td>G</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gid</td>\n",
       "      <td>Rv3919c</td>\n",
       "      <td>352</td>\n",
       "      <td>del</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>118</td>\n",
       "      <td>...</td>\n",
       "      <td>21.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>115.0</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26713</th>\n",
       "      <td>2288850</td>\n",
       "      <td>A</td>\n",
       "      <td>ACC</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>pncA</td>\n",
       "      <td>Rv2043c</td>\n",
       "      <td>392</td>\n",
       "      <td>ins</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>131</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50146</th>\n",
       "      <td>4407851</td>\n",
       "      <td>G</td>\n",
       "      <td>GC</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gid</td>\n",
       "      <td>Rv3919c</td>\n",
       "      <td>352</td>\n",
       "      <td>ins</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>118</td>\n",
       "      <td>...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           pos ref  alt          gene_category gene_name  gene_id gene_pos  \\\n",
       "50204  4408100  GC    G  Antibiotic Resistance       gid  Rv3919c      103   \n",
       "50198  4408087  CG    C  Antibiotic Resistance       gid  Rv3919c      116   \n",
       "50144  4407851  GC    G  Antibiotic Resistance       gid  Rv3919c      352   \n",
       "26713  2288850   A  ACC  Antibiotic Resistance      pncA  Rv2043c      392   \n",
       "50146  4407851   G   GC  Antibiotic Resistance       gid  Rv3919c      352   \n",
       "\n",
       "      ins_del  INDEL_type codon_pos      ...          1     2     3    4A  \\\n",
       "50204     del  frameshift        35      ...       41.0  27.0  51.0  43.0   \n",
       "50198     del  frameshift        39      ...       23.0  11.0  35.0  33.0   \n",
       "50144     del  frameshift       118      ...       21.0   9.0  40.0  20.0   \n",
       "26713     ins  frameshift       131      ...        1.0  14.0   6.0   3.0   \n",
       "50146     ins  frameshift       118      ...        5.0   1.0  12.0   5.0   \n",
       "\n",
       "         4B    4C    5    6  total  num_isolates  \n",
       "50204  11.0  29.0  0.0  0.0  202.0           388  \n",
       "50198  10.0  18.0  0.0  0.0  130.0           212  \n",
       "50144   6.0  19.0  0.0  0.0  115.0           182  \n",
       "26713   6.0   2.0  0.0  0.0   32.0            48  \n",
       "50146   7.0   1.0  0.0  0.0   31.0            52  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_homoplasic_INDELs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(46306, 22)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(all_homoplasic_INDELs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "create **key** column for INDELs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_homoplasic_INDELs_key_col = [ref + '_' + str(int(pos)) + '_' + alt for ref,pos,alt in zip(all_homoplasic_INDELs.ref,all_homoplasic_INDELs.pos,all_homoplasic_INDELs.alt)]\n",
    "all_homoplasic_INDELs.loc[:,'key'] = all_homoplasic_INDELs_key_col"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_pos</th>\n",
       "      <th>ins_del</th>\n",
       "      <th>INDEL_type</th>\n",
       "      <th>codon_pos</th>\n",
       "      <th>...</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4A</th>\n",
       "      <th>4B</th>\n",
       "      <th>4C</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>total</th>\n",
       "      <th>num_isolates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50204</th>\n",
       "      <td>4408100</td>\n",
       "      <td>GC</td>\n",
       "      <td>G</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gid</td>\n",
       "      <td>Rv3919c</td>\n",
       "      <td>103</td>\n",
       "      <td>del</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>35</td>\n",
       "      <td>...</td>\n",
       "      <td>41.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>202.0</td>\n",
       "      <td>388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50198</th>\n",
       "      <td>4408087</td>\n",
       "      <td>CG</td>\n",
       "      <td>C</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gid</td>\n",
       "      <td>Rv3919c</td>\n",
       "      <td>116</td>\n",
       "      <td>del</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>39</td>\n",
       "      <td>...</td>\n",
       "      <td>23.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>130.0</td>\n",
       "      <td>212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50144</th>\n",
       "      <td>4407851</td>\n",
       "      <td>GC</td>\n",
       "      <td>G</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gid</td>\n",
       "      <td>Rv3919c</td>\n",
       "      <td>352</td>\n",
       "      <td>del</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>118</td>\n",
       "      <td>...</td>\n",
       "      <td>21.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>115.0</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26713</th>\n",
       "      <td>2288850</td>\n",
       "      <td>A</td>\n",
       "      <td>ACC</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>pncA</td>\n",
       "      <td>Rv2043c</td>\n",
       "      <td>392</td>\n",
       "      <td>ins</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>131</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50146</th>\n",
       "      <td>4407851</td>\n",
       "      <td>G</td>\n",
       "      <td>GC</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gid</td>\n",
       "      <td>Rv3919c</td>\n",
       "      <td>352</td>\n",
       "      <td>ins</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>118</td>\n",
       "      <td>...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           pos ref  alt          gene_category gene_name  gene_id gene_pos  \\\n",
       "50204  4408100  GC    G  Antibiotic Resistance       gid  Rv3919c      103   \n",
       "50198  4408087  CG    C  Antibiotic Resistance       gid  Rv3919c      116   \n",
       "50144  4407851  GC    G  Antibiotic Resistance       gid  Rv3919c      352   \n",
       "26713  2288850   A  ACC  Antibiotic Resistance      pncA  Rv2043c      392   \n",
       "50146  4407851   G   GC  Antibiotic Resistance       gid  Rv3919c      352   \n",
       "\n",
       "      ins_del  INDEL_type codon_pos      ...          1     2     3    4A  \\\n",
       "50204     del  frameshift        35      ...       41.0  27.0  51.0  43.0   \n",
       "50198     del  frameshift        39      ...       23.0  11.0  35.0  33.0   \n",
       "50144     del  frameshift       118      ...       21.0   9.0  40.0  20.0   \n",
       "26713     ins  frameshift       131      ...        1.0  14.0   6.0   3.0   \n",
       "50146     ins  frameshift       118      ...        5.0   1.0  12.0   5.0   \n",
       "\n",
       "         4B    4C    5    6  total  num_isolates  \n",
       "50204  11.0  29.0  0.0  0.0  202.0           388  \n",
       "50198  10.0  18.0  0.0  0.0  130.0           212  \n",
       "50144   6.0  19.0  0.0  0.0  115.0           182  \n",
       "26713   6.0   2.0  0.0  0.0   32.0            48  \n",
       "50146   7.0   1.0  0.0  0.0   31.0            52  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_homoplasic_INDELs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "INDEL_variant_keys = set(all_homoplasic_INDELs.key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "46306"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(INDEL_variant_keys)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note - SNPs & INDELs with homoplasy score $ = 0$ were dropped from DataFrames"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# [2] Load in mutations detected in *mmpR*, *mmpL5*, *mmpS5*, *ahpC*, *whiB7*, *eis* and upstream regions for *ahpC*, *whiB7*, *eis*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These mutations are taken from **Supplementary Table 1** after it has been created."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations_of_interest_df = pd.read_excel('/n/data1/hms/dbmi/farhat/Roger/mmpR_BDQ_mutant_project/CSV files/mutations_detected_within_isolates/Supplementary Table 1.xlsx')\n",
    "\n",
    "#drop unneccessary cols\n",
    "mutations_of_interest_df.drop(labels=['relative to start', 'relative to TSS', '*Cepheid Xpert MTB/XDR Assay'], axis=1, inplace = True)\n",
    "\n",
    "#drop unneccessary rows\n",
    "mutations_of_interest_df.drop(labels=mutations_of_interest_df.loc[334:,:].index, axis=0, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Allele</th>\n",
       "      <th>Alternate Allele</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>H37Rv locus tag</th>\n",
       "      <th>Gene Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Codon Position</th>\n",
       "      <th># Isolates</th>\n",
       "      <th># Sublineages</th>\n",
       "      <th>sublineages</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>775616</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>mmpL5</td>\n",
       "      <td>Rv0676c</td>\n",
       "      <td>2865</td>\n",
       "      <td>ins</td>\n",
       "      <td>955</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.13(1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>775645</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>mmpL5</td>\n",
       "      <td>Rv0676c</td>\n",
       "      <td>2836</td>\n",
       "      <td>N</td>\n",
       "      <td>R946*</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.2.1.1.1(1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>775727</td>\n",
       "      <td>GGT</td>\n",
       "      <td>G</td>\n",
       "      <td>mmpL5</td>\n",
       "      <td>Rv0676c</td>\n",
       "      <td>2754</td>\n",
       "      <td>del</td>\n",
       "      <td>918</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.2.1.1.1(3)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>775753</td>\n",
       "      <td>TG</td>\n",
       "      <td>T</td>\n",
       "      <td>mmpL5</td>\n",
       "      <td>Rv0676c</td>\n",
       "      <td>2728</td>\n",
       "      <td>del</td>\n",
       "      <td>910</td>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5(14)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>775876</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>mmpL5</td>\n",
       "      <td>Rv0676c</td>\n",
       "      <td>2605</td>\n",
       "      <td>ins</td>\n",
       "      <td>869</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.1.1.1(2)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Position Reference Allele Alternate Allele Gene Symbol H37Rv locus tag  \\\n",
       "0   775616                C               CG       mmpL5         Rv0676c   \n",
       "1   775645                G                A       mmpL5         Rv0676c   \n",
       "2   775727              GGT                G       mmpL5         Rv0676c   \n",
       "3   775753               TG                T       mmpL5         Rv0676c   \n",
       "4   775876                C               CG       mmpL5         Rv0676c   \n",
       "\n",
       "  Gene Position Mutation Type Codon Position  # Isolates  # Sublineages  \\\n",
       "0          2865           ins            955         1.0            1.0   \n",
       "1          2836             N          R946*         1.0            1.0   \n",
       "2          2754           del            918         3.0            1.0   \n",
       "3          2728           del            910        14.0            1.0   \n",
       "4          2605           ins            869         2.0            1.0   \n",
       "\n",
       "    sublineages  \n",
       "0       4.13(1)  \n",
       "1  2.2.1.1.1(1)  \n",
       "2  2.2.1.1.1(3)  \n",
       "3         5(14)  \n",
       "4    1.1.1.1(2)  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations_of_interest_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Allele</th>\n",
       "      <th>Alternate Allele</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>H37Rv locus tag</th>\n",
       "      <th>Gene Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Codon Position</th>\n",
       "      <th># Isolates</th>\n",
       "      <th># Sublineages</th>\n",
       "      <th>sublineages</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>329</th>\n",
       "      <td>3569048</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>whiB7-uvrD2</td>\n",
       "      <td>Rv3197A-Rv3198c</td>\n",
       "      <td>None</td>\n",
       "      <td>I</td>\n",
       "      <td>None</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.1.2(1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330</th>\n",
       "      <td>3569058</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>whiB7-uvrD2</td>\n",
       "      <td>Rv3197A-Rv3198c</td>\n",
       "      <td>None</td>\n",
       "      <td>I</td>\n",
       "      <td>None</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.10.i1(4)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>331</th>\n",
       "      <td>3569065</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>whiB7-uvrD2</td>\n",
       "      <td>Rv3197A-Rv3198c</td>\n",
       "      <td>None</td>\n",
       "      <td>I</td>\n",
       "      <td>None</td>\n",
       "      <td>66.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.3.i1(49) - 3.1.1(1) - 4.2.1.1.1.2.2.1(15) - ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>332</th>\n",
       "      <td>3569069</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>whiB7-uvrD2</td>\n",
       "      <td>Rv3197A-Rv3198c</td>\n",
       "      <td>None</td>\n",
       "      <td>I</td>\n",
       "      <td>None</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.1.1(1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>333</th>\n",
       "      <td>3569081</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>whiB7-uvrD2</td>\n",
       "      <td>Rv3197A-Rv3198c</td>\n",
       "      <td>None</td>\n",
       "      <td>I</td>\n",
       "      <td>None</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.1.2(2)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Position Reference Allele Alternate Allele  Gene Symbol  H37Rv locus tag  \\\n",
       "329  3569048                G                A  whiB7-uvrD2  Rv3197A-Rv3198c   \n",
       "330  3569058                G                A  whiB7-uvrD2  Rv3197A-Rv3198c   \n",
       "331  3569065                A                G  whiB7-uvrD2  Rv3197A-Rv3198c   \n",
       "332  3569069                C                A  whiB7-uvrD2  Rv3197A-Rv3198c   \n",
       "333  3569081                C                A  whiB7-uvrD2  Rv3197A-Rv3198c   \n",
       "\n",
       "    Gene Position Mutation Type Codon Position  # Isolates  # Sublineages  \\\n",
       "329          None             I           None         1.0            1.0   \n",
       "330          None             I           None         4.0            1.0   \n",
       "331          None             I           None        66.0            4.0   \n",
       "332          None             I           None         1.0            1.0   \n",
       "333          None             I           None         2.0            1.0   \n",
       "\n",
       "                                           sublineages  \n",
       "329                                           1.1.2(1)  \n",
       "330                                         4.10.i1(4)  \n",
       "331  4.3.i1(49) - 3.1.1(1) - 4.2.1.1.1.2.2.1(15) - ...  \n",
       "332                                           3.1.1(1)  \n",
       "333                                           1.1.2(2)  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations_of_interest_df.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "create **key** column for variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations_of_interest_df_key_col = [ref + '_' + str(int(pos)) + '_' + alt for ref,pos,alt in zip(mutations_of_interest_df.loc[:,'Reference Allele'],mutations_of_interest_df.loc[:,'Position'],mutations_of_interest_df.loc[:,'Alternate Allele'])]\n",
    "mutations_of_interest_df.loc[:,'key'] = mutations_of_interest_df_key_col"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Allele</th>\n",
       "      <th>Alternate Allele</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>H37Rv locus tag</th>\n",
       "      <th>Gene Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Codon Position</th>\n",
       "      <th># Isolates</th>\n",
       "      <th># Sublineages</th>\n",
       "      <th>sublineages</th>\n",
       "      <th>key</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>775616</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>mmpL5</td>\n",
       "      <td>Rv0676c</td>\n",
       "      <td>2865</td>\n",
       "      <td>ins</td>\n",
       "      <td>955</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.13(1)</td>\n",
       "      <td>C_775616_CG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>775645</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>mmpL5</td>\n",
       "      <td>Rv0676c</td>\n",
       "      <td>2836</td>\n",
       "      <td>N</td>\n",
       "      <td>R946*</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.2.1.1.1(1)</td>\n",
       "      <td>G_775645_A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>775727</td>\n",
       "      <td>GGT</td>\n",
       "      <td>G</td>\n",
       "      <td>mmpL5</td>\n",
       "      <td>Rv0676c</td>\n",
       "      <td>2754</td>\n",
       "      <td>del</td>\n",
       "      <td>918</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.2.1.1.1(3)</td>\n",
       "      <td>GGT_775727_G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>775753</td>\n",
       "      <td>TG</td>\n",
       "      <td>T</td>\n",
       "      <td>mmpL5</td>\n",
       "      <td>Rv0676c</td>\n",
       "      <td>2728</td>\n",
       "      <td>del</td>\n",
       "      <td>910</td>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5(14)</td>\n",
       "      <td>TG_775753_T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>775876</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>mmpL5</td>\n",
       "      <td>Rv0676c</td>\n",
       "      <td>2605</td>\n",
       "      <td>ins</td>\n",
       "      <td>869</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.1.1.1(2)</td>\n",
       "      <td>C_775876_CG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Position Reference Allele Alternate Allele Gene Symbol H37Rv locus tag  \\\n",
       "0   775616                C               CG       mmpL5         Rv0676c   \n",
       "1   775645                G                A       mmpL5         Rv0676c   \n",
       "2   775727              GGT                G       mmpL5         Rv0676c   \n",
       "3   775753               TG                T       mmpL5         Rv0676c   \n",
       "4   775876                C               CG       mmpL5         Rv0676c   \n",
       "\n",
       "  Gene Position Mutation Type Codon Position  # Isolates  # Sublineages  \\\n",
       "0          2865           ins            955         1.0            1.0   \n",
       "1          2836             N          R946*         1.0            1.0   \n",
       "2          2754           del            918         3.0            1.0   \n",
       "3          2728           del            910        14.0            1.0   \n",
       "4          2605           ins            869         2.0            1.0   \n",
       "\n",
       "    sublineages           key  \n",
       "0       4.13(1)   C_775616_CG  \n",
       "1  2.2.1.1.1(1)    G_775645_A  \n",
       "2  2.2.1.1.1(3)  GGT_775727_G  \n",
       "3         5(14)   TG_775753_T  \n",
       "4    1.1.1.1(2)   C_775876_CG  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations_of_interest_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(334, 12)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(mutations_of_interest_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# [3] Merge homoplasy counts to DataFrame with mutations of interest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "#iterate through each mutation and find the homoplasy count\n",
    "homoplasy_count_list = []\n",
    "\n",
    "for mut_key in mutations_of_interest_df.key:\n",
    "    \n",
    "    if mut_key in SNP_variant_keys:\n",
    "        homoplasy_count_list.append(int(all_homoplasic_SNPs[all_homoplasic_SNPs.key == mut_key].total))\n",
    "        \n",
    "    elif mut_key in INDEL_variant_keys:\n",
    "        homoplasy_count_list.append(int(all_homoplasic_INDELs[all_homoplasic_INDELs.key == mut_key].total))\n",
    "        \n",
    "    else:\n",
    "        homoplasy_count_list.append(0) #SNPs & INDELs with homoplasy score = 0 were dropped from DataFrames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations_of_interest_df.loc[:,'Homoplasy Score'] = homoplasy_count_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference Allele</th>\n",
       "      <th>Alternate Allele</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>H37Rv locus tag</th>\n",
       "      <th>Gene Position</th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Codon Position</th>\n",
       "      <th># Isolates</th>\n",
       "      <th># Sublineages</th>\n",
       "      <th>sublineages</th>\n",
       "      <th>key</th>\n",
       "      <th>Homoplasy Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>775616</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>mmpL5</td>\n",
       "      <td>Rv0676c</td>\n",
       "      <td>2865</td>\n",
       "      <td>ins</td>\n",
       "      <td>955</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.13(1)</td>\n",
       "      <td>C_775616_CG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>775645</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>mmpL5</td>\n",
       "      <td>Rv0676c</td>\n",
       "      <td>2836</td>\n",
       "      <td>N</td>\n",
       "      <td>R946*</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.2.1.1.1(1)</td>\n",
       "      <td>G_775645_A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>775727</td>\n",
       "      <td>GGT</td>\n",
       "      <td>G</td>\n",
       "      <td>mmpL5</td>\n",
       "      <td>Rv0676c</td>\n",
       "      <td>2754</td>\n",
       "      <td>del</td>\n",
       "      <td>918</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.2.1.1.1(3)</td>\n",
       "      <td>GGT_775727_G</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>775753</td>\n",
       "      <td>TG</td>\n",
       "      <td>T</td>\n",
       "      <td>mmpL5</td>\n",
       "      <td>Rv0676c</td>\n",
       "      <td>2728</td>\n",
       "      <td>del</td>\n",
       "      <td>910</td>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5(14)</td>\n",
       "      <td>TG_775753_T</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>775876</td>\n",
       "      <td>C</td>\n",
       "      <td>CG</td>\n",
       "      <td>mmpL5</td>\n",
       "      <td>Rv0676c</td>\n",
       "      <td>2605</td>\n",
       "      <td>ins</td>\n",
       "      <td>869</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.1.1.1(2)</td>\n",
       "      <td>C_775876_CG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Position Reference Allele Alternate Allele Gene Symbol H37Rv locus tag  \\\n",
       "0   775616                C               CG       mmpL5         Rv0676c   \n",
       "1   775645                G                A       mmpL5         Rv0676c   \n",
       "2   775727              GGT                G       mmpL5         Rv0676c   \n",
       "3   775753               TG                T       mmpL5         Rv0676c   \n",
       "4   775876                C               CG       mmpL5         Rv0676c   \n",
       "\n",
       "  Gene Position Mutation Type Codon Position  # Isolates  # Sublineages  \\\n",
       "0          2865           ins            955         1.0            1.0   \n",
       "1          2836             N          R946*         1.0            1.0   \n",
       "2          2754           del            918         3.0            1.0   \n",
       "3          2728           del            910        14.0            1.0   \n",
       "4          2605           ins            869         2.0            1.0   \n",
       "\n",
       "    sublineages           key  Homoplasy Score  \n",
       "0       4.13(1)   C_775616_CG                1  \n",
       "1  2.2.1.1.1(1)    G_775645_A                1  \n",
       "2  2.2.1.1.1(3)  GGT_775727_G                1  \n",
       "3         5(14)   TG_775753_T                1  \n",
       "4    1.1.1.1(2)   C_775876_CG                1  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations_of_interest_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.collections.PathCollection at 0x7f0896820810>"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAD8CAYAAAB5Pm/hAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAEu5JREFUeJzt3WuMXHd5x/HvExtDllASkyW4dlKHyjRQ1HJZRVAQsqAsxEWTVMJVtlR1WySnkt1CLyoQXiRvkkJLDa2yjeSSgInS5WJMPapoO1EaBH1ByjqgxIkhiSCETUy8Ubgk3YAJefpiz4bFjHd3bjuz//l+pNXMOXPOzHN85N+efeac84/MRJJUrjP6XYAkqbcMekkqnEEvSYUz6CWpcAa9JBXOoJekwhn0klQ4g16SCmfQS1Lh1ve7AIBzzz03t27d2u8yJGlNOXLkyKOZObrccgMR9Fu3bmV6errfZUjSmhIR317JcrZuJKlwBr0kFc6gl6TCGfSSVDiDXpIKZ9BLUuEMeklN1et19u7dS71e73cp6pBBL+kX1Ot1JiYmmJycZGJiwrBf4wx6Sb+g0WgwNzcHwNzcHI1Go88VqRMGvaRfMD4+zsjICAAjIyOMj4/3uSJ1YiBugSBpsNRqNaampmg0GoyPj1Or1fpdkjoQmdnvGhgbG0vvdSNJrYmII5k5ttxytm4kqXAGvSQVzqCXpMIZ9JJUOINekgpn0EtS4Qx6SSqcQS9JhTPoJalwywZ9RNwYESci4uiieX8fEV+PiDsj4nMRcfai194XEfdHxDci4i29KlyStDIrOaL/OPDWU+bdArw8M38DuBd4H0BEvAy4HPj1ap1/joh1XatWktSyZYM+M78IPHbKvEZmPlVNfhnYUj2/FPhkZv44M78F3A9c3MV6JUkt6kaP/k+A/6iebwa+s+i1mWqeJKlPOgr6iHg/8BRw88KsJos1vT1mROyOiOmImJ6dne2kDEnSEtoO+ojYBbwNeEf+7F7HM8D5ixbbAjzcbP3M3J+ZY5k5Njo62m4ZkqRltBX0EfFW4D1ALTPnFr1UBy6PiGdHxIXANuB/Oy9TktSuZUeYiogpYDtwbkTMAFcxf5bNs4FbIgLgy5n5p5l5d0R8GriH+ZbOnsz8aa+KlyQtzxGmJGmNcoQpSRJg0EtS8Qx6SSqcQS9JhTPoJalwBr0kFc6gl6TCGfSSVDiDXpIKZ9BLUuEMekkqnEEvSYUz6CWpcAa9JBXOoJekwhn0klQ4g16SCmfQS1LhDHpJKpxBL0mFM+glqXDLBn1E3BgRJyLi6KJ5GyPiloi4r3o8p5ofEfFPEXF/RNwZEa/qZfGSpOWt5Ij+48BbT5n3XuDWzNwG3FpNA1wCbKt+dgPXd6dMSVK7lg36zPwi8Ngpsy8FDlTPDwCXLZr/iZz3ZeDsiNjUrWIlSa1rt0d/XmYeB6geX1jN3wx8Z9FyM9U8SVKfdPvL2GgyL5suGLE7IqYjYnp2drbLZUiSFrQb9I8stGSqxxPV/Bng/EXLbQEebvYGmbk/M8cyc2x0dLTNMiRJy2k36OvArur5LuDwovl/WJ198xrgBwstHklSf6xfboGImAK2A+dGxAxwFfAB4NMR8U7gQWBntfjngR3A/cAc8Mc9qFmS1IJlgz4zJ07z0puaLJvAnk6LkiR1j1fGSlLhDHpJKpxBL0mFM+glqXAGvSQVzqCXpMIZ9JJUOINekgpn0EtS4Qx6SSqcQS9JhTPoJalwBr0kFc6gl6TCGfSSVDiDXpIKZ9BLUuEMekkqnEEvSYUz6CWpcAa9pK6p1+vs3buXer3e71K0iEEvqSvq9ToTExNMTk4yMTFh2A+QjoI+Iv4iIu6OiKMRMRURz4mICyPi9oi4LyI+FREbulWspMHVaDSYm5sDYG5ujkaj0eeKtKDtoI+IzcCfA2OZ+XJgHXA58EHgw5m5Dfge8M5uFCppsI2PjzMyMgLAyMgI4+Pjfa5IC9Z3Yf0zI+InwAhwHHgj8PvV6weAq4HrO/wcSQOuVqsxNTVFo9FgfHycWq3W75JUaTvoM/OhiPgQ8CDwJNAAjgDfz8ynqsVmgM3N1o+I3cBugAsuuKDdMiQNkFqtZsAPoE5aN+cAlwIXAr8MPBe4pMmi2Wz9zNyfmWOZOTY6OtpuGZKkZXTyZexvA9/KzNnM/AlwCPgt4OyIWPhLYQvwcIc1SpI60EnQPwi8JiJGIiKANwH3ALcBb6+W2QUc7qxESVIn2g76zLwdOAjcAdxVvdd+4D3AX0bE/cALgBu6UKckqU0dnXWTmVcBV50y+5vAxZ28rySpe7wyVhoC3ppguBn0UuG8NYEMeqlw3ppABr1UOG9NoE5vgSBpwHlrAkVm0wtXV9XY2FhOT0/3uwxJWlMi4khmji23nK0bSSqcQS9JhTPoJalwBr0kFc6gl6TCGfSSVDiDXpIKZ9BLUuEMekkqnEEvSYUz6CWpcAa9JBXOoJekwhn0klQ4g16SCtdR0EfE2RFxMCK+HhHHIuK1EbExIm6JiPuqx3O6VawkqXWdHtH/I/CfmXkR8JvAMeC9wK2ZuQ24tZqWJPVJ20EfEb8EvAG4ASAzT2bm94FLgQPVYgeAyzotUpLUvk6O6F8MzAIfi4ivRsRHI+K5wHmZeRygenxhs5UjYndETEfE9OzsbAdlSJKW0knQrwdeBVyfma8E/o8W2jSZuT8zxzJzbHR0tIMyJGltqtfr7N27l3q93tPP6SToZ4CZzLy9mj7IfPA/EhGbAKrHE52VKEnlqdfrTExMMDk5ycTERE/Dvu2gz8zvAt+JiF+rZr0JuAeoA7uqebuAwx1VKEkFajQazM3NATA3N0ej0ejZZ3V61s2fATdHxJ3AK4BrgQ8Ab46I+4A3V9OSpEXGx8cZGRkBYGRkhPHx8Z59VmRmz958pcbGxnJ6errfZUjSqqrX6zQaDcbHx6nVai2vHxFHMnNs2eUMeklam1Ya9N4CQZIKZ9BLUuEMekl9s1rnkQ87g15SX6zmeeTDzqCX1BereR75sDPoJfXFap5HPuzW97sAScOpVqsxNTXV0XnkWhnPo5ekNcrz6CVJgEEvScUz6CWpcAa9JBXOoJekwhn0klQ4g16SCmfQS1LhDHpJKpxBL0mFM+glqXAGvSQVruOgj4h1EfHViPj3avrCiLg9Iu6LiE9FxIbOy5QktasbR/TvAo4tmv4g8OHM3AZ8D3hnFz5DktSmjoI+IrYAvwN8tJoO4I3AwWqRA8BlnXyG+ssxPaW1r9Mj+o8AfwM8XU2/APh+Zj5VTc8Amzv8DPWJY3oOLn8BqxVtB31EvA04kZlHFs9usmjTkU0iYndETEfE9OzsbLtlqIcc03Mw+QtYrerkiP51QC0iHgA+yXzL5iPA2RGxMEThFuDhZitn5v7MHMvMsdHR0Q7KUK84pudg8hewWtV20Gfm+zJzS2ZuBS4H/jsz3wHcBry9WmwXcLjjKtUXC2N67tmzh6mpKcf0HBDD/gvYtlXrujJmbERsB/46M98WES9m/gh/I/BV4A8y88dLre+YsVJr6vX6UA6qvdC2mpubY2RkZOgPQFY6Zuz65RZYicz8AvCF6vk3gYu78b6SmqvVakMZcM3aVsP479Aqr4yVtGYMe9uqXV05opek1bDwvdEwtq060ZUefafs0UtS61bao7d1I0mFM+glqXAGvSQVzqDXmuRFM9LKGfRac7zXi9Qag15rjvd6kVpj0GvN8aIZqTVeMKU1x4tmpNZ4wZQkrVFeMCVJAgx6SSqeQS9JhTPopSWsxoVZXvylXvPLWOk0VmM0I0dMUif8Mlbq0GpcmOXFX1oNBr10GqtxYZYXf2k12LqRlrAag3AP60Df6txKWzcGvSStUT3v0UfE+RFxW0Qci4i7I+Jd1fyNEXFLRNxXPZ7T7mdIkjrXSY/+KeCvMvOlwGuAPRHxMuC9wK2ZuQ24tZqWJPVJ20Gfmccz847q+ePAMWAzcClwoFrsAHBZp0VKktrXlbNuImIr8ErgduC8zDwO878MgBd24zMkSe3pOOgj4izgs8C7M/OHLay3OyKmI2J6dna20zIkSafRUdBHxLOYD/mbM/NQNfuRiNhUvb4JONFs3czcn5ljmTk2OjraSRmSpCV0ctZNADcAxzJz36KX6sCu6vku4HD75WlYeL8XqXfaPo8+Il4PfAm4C3i6mn0l8336TwMXAA8COzPzsaXey/Poh5v3e5Has9Lz6NseSjAz/weI07z8pnbfV8On2f1eDHqpe4buXje2CAaP93uRemuoboFgi2Bweb8XqXU9b92sRbYIBletVnNfSD0yVK0bWwSts9UlrX1D1boBWwStsNUlDTZHmDqNWq3Gdddd11JgDepRba/rcvQjqQxDF/StWjiqnZycZGJiYmDCfjXqstUllcGgX8agHtWuRl21Wo2pqSn27Nlj20Zawwz6ZQzqUW27dbXa7mmn1SVpsAzdl7HtaOcL3EEca9QvV6WyOGZsHw1qoO7du5fJyclnpvfs2cN1113Xx4okdcKzbvqonf75apzZM6htKEm9taaDvp1wXI1Aff7zn7/kdLOadu7cyeTkJDt37uxZbX65Kg2pzOz7z6tf/eps1eHDh3NkZCSBHBkZycOHD/dknXbs2LEjgWd+duzY0dXlJSkzE5jOFWTsmj2ib6c9MqinSkpSL63ZoG+n39xqS2VBq+2eK664gg0bNgCwYcMGrrjiimWXX79+/v5y69evX3Z5SWrFmr175UK/uZXTC3/wgx8sOd3M4jNoPvaxj62ot12r1fjMZz7TUm1nnHHGzz1KUres6VS56aabmJqa4qabblrR8u38FdBuu6eV2hqNBidPngTg5MmTtpQkddWaDfqdO3dy8OBBHnvsMQ4ePMjOnTuXXadWq7Fjxw42btzIjh07VnSk3U67p9XaPO1RUi+t2QumzjzzTH70ox89M/2c5zyHJ598csl13v/+93Pttdc+M33llVdyzTXXLLnOpk2b+O53v/vM9Ite9CKOHz++5DrPe97zeOKJJ56ZPuuss3j88ceXXMfbJ0tqVfEXTC182Xm66WZO/TJ1JV+uPvroo0tON3PeeectOd2M95TpvUG93bTUaz0L+oh4a0R8IyLuj4j3dvv9F3rap5tu5ujRo0tON3PxxRcvOd3Mvn37WLduHQDr1q1j3759y66j3hrU201Lq6EnQR8R64BJ4BLgZcBERLysm5+xuG3TbLpbtm/fvuR0M7VajUOHDrFnzx4OHTrkUfoA8BoKDbNeHdFfDNyfmd/MzJPAJ4FLe/RZPdVOuwdsxQwav/DWMOtV0G8GvrNoeqaa11cXXXTRktPNnBrUBvfa5H1+NMx6dcFUNJn3c6f3RMRuYDfABRdc0KMyft6xY8d46Utfyr333stLXvISjh07tuw6C2fl1Ot1arXasmfpaHDVajUDXkOpJ6dXRsRrgasz8y3V9PsAMvNvmy3fzumVW7Zs4aGHHnpmevPmzczMzLRdsyStNf0+vfIrwLaIuDAiNgCXA109zWFmZobNm+e7QYa8JJ1eT1o3mflUROwF/gtYB9yYmXd3+3MMd0laXs9uapaZnwc+36v3lyStzJq9MlaStDIGvSQVzqCXpMIZ9JJUOINekgo3EPejj4hZ4NvV5LnA8vcCLtMwbzsM9/a77cOrk+3/lcwcXW6hgQj6xSJieiVXepVomLcdhnv73fbh3HZYne23dSNJhTPoJalwgxj0+/tdQB8N87bDcG+/2z68er79A9ejlyR11yAe0UuSumhggr7Xg4kPuoh4ICLuioivRURrN+dfYyLixog4ERFHF83bGBG3RMR91eM5/ayxl06z/VdHxEPV/v9aROzoZ429EhHnR8RtEXEsIu6OiHdV84vf/0tse8/3/UC0bqrBxO8F3sz8sINfASYy856+FraKIuIBYCwziz+fOCLeADwBfCIzX17N+zvgscz8QPWL/pzMfE8/6+yV02z/1cATmfmhftbWaxGxCdiUmXdExPOAI8BlwB9R+P5fYtt/jx7v+0E5oi9mMHEtLzO/CDx2yuxLgQPV8wPM/wco0mm2fyhk5vHMvKN6/jhwjPnxpIvf/0tse88NStAP5GDiqyyBRkQcqcbTHTbnZeZxmP8PAbywz/X0w96IuLNq7RTXujhVRGwFXgnczpDt/1O2HXq87wcl6JcdTHwIvC4zXwVcAuyp/rzX8Lge+FXgFcBx4B/6W05vRcRZwGeBd2fmD/tdz2pqsu093/eDEvQzwPmLprcAD/eplr7IzIerxxPA55hvZw2TR6oe5kIv80Sf61lVmflIZv40M58G/oWC939EPIv5oLs5Mw9Vs4di/zfb9tXY94MS9D0fTHyQRcRzqy9niIjnAuPA0aXXKk4d2FU93wUc7mMtq24h5Cq/S6H7PyICuAE4lpn7Fr1U/P4/3bavxr4fiLNuAKpTij7CzwYTv6bPJa2aiHgx80fxMD+O77+WvP0RMQVsZ/6ufY8AVwH/BnwauAB4ENiZmUV+YXma7d/O/J/uCTwAXLHQsy5JRLwe+BJwF/B0NftK5nvVRe//JbZ9gh7v+4EJeklSbwxK60aS1CMGvSQVzqCXpMIZ9JJUOINekgpn0EtS4Qx6SSqcQS9Jhft/n29da6Wwl8EAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.scatter(mutations_of_interest_df.loc[:, '# Sublineages'], mutations_of_interest_df.loc[:, 'Homoplasy Score'], s = 10, c = 'k')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Output to copy/paste **Homplasy Score** column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations_of_interest_df.to_excel('/n/data1/hms/dbmi/farhat/Roger/mmpR_BDQ_mutant_project/CSV files/mutations_detected_within_isolates/Supplementary Table 1 with HP scores.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Infer whether a variant is homoplasic by whether it is present in multiple **sub-lineages** or from **homoplasy score**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "homoplasic_from_sublineage_count_boolean = list(mutations_of_interest_df.loc[:,'# Sublineages'] > 1)\n",
    "homoplasic_from_homoplasy_score_boolean = list(mutations_of_interest_df.loc[:,'Homoplasy Score'] > 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "73"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(homoplasic_from_sublineage_count_boolean)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "79"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(homoplasic_from_homoplasy_score_boolean)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "66"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum([(homoplasic_from_sublineage_count and homoplasic_from_homoplasy_score) for homoplasic_from_sublineage_count,homoplasic_from_homoplasy_score in zip(homoplasic_from_sublineage_count_boolean,homoplasic_from_homoplasy_score_boolean)])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
